Expose bytecode to Python

tugsbayasgalan opened the pull request:
In this PR, we expose upgrader bytecode to mobile through pybind. 
1. We convert TS graph of upgrader to GraphFunction
2. We call compileJitFunctionToMobileFunction to convert GraphFunction to mobile function.
3. We convert the mobile function to a code table of IValues (note that this should be changed in the future when we migrate away from Ivalue based storage) 
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68773
* #68772
* __->__ #66592
* #66591

Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

tugsbayasgalan committed:
Expose MobileCode to python

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Expose MobileCode to python"


Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Expose MobileCode to python"


Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632600).

tugsbayasgalan committed:
Update on "Expose MobileCode to python"


Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Expose MobileCode to python"


Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632600).

cccclai commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -248,7 +252,54 @@ std::unique_ptr<mobile::Code> compileGraphToMobileCode(
+  }
+
+  std::vector<IValue> types;
+  for (const TypePtr& t : code->types_) {
```
Probably leave a comment here, explaining NameTuple is not supported in upgrader bytecode at the moment. Ideally it's better to throw an exception if we see a NamedTuple here.

qihqi commented on torch/_C/__init__.pyi.in:
```diff
@@ -189,6 +189,8 @@ def _jit_pass_convert_frozen_ops_to_mkldnn(graph: Graph): ...
 def _jit_pass_transpose_frozen_linear(graph:Graph): ...
 def _jit_pass_remove_dropout(module: 'torch.jit.ScriptModule'): ...
 
+def _compile_graph_to_code_table(name: str, graph: Graph, compilation_options: CompilationOptions) -> IValue: ...
```
remove compilation_options? as in script_init you didnt define it

tugsbayasgalan replied:
good catch!

tugsbayasgalan replied:
Yep good catch!

qihqi commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1593,6 +1599,16 @@ void initJitScriptBindings(PyObject* module) {
+  m.def("_compile_graph_to_code_table",
+        [](
+          const std::string& name,
+          const std::shared_ptr<Graph>& graph) {
```
here it doesnt have compilation options as declared in the .pyi.in file

tugsbayasgalan replied:
Yep good catch!

qihqi commented on torch/csrc/jit/runtime/instruction.h:
```diff
@@ -63,6 +63,8 @@ enum OpCode : uint8_t {
 #undef DEFINE_OP
 };
 
+TORCH_API char const* toString(OpCode op);
```
this is declared in line 80 already

qihqi commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -137,7 +137,7 @@ std::unique_ptr<mobile::Code> compileGraphToMobileCode(
     const std::shared_ptr<Graph>& graph,
     const CompilationOptions& compilation_options,
-    BackendDebugInfoRecorder& debug_info_recorder) {
+    BackendDebugInfoRecorder* debug_info_recorder) {
```
Don't make this optional.

The caller can already make it optional by passing one and not using.

tugsbayasgalan replied:
Hmm do you mean just initialize BackendDebugInfoRecorder in the script_init?

qihqi replied:
yeah just declare a new one with `BackendDebugInfoRecorder recorder` and pass it along.

qihqi commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -248,7 +252,54 @@ std::unique_ptr<mobile::Code> compileGraphToMobileCode(
 }
 
-void checkSchema(const FunctionSchema& schema) {
+std::unique_ptr<mobile::Function> convertJitFunctionToMobileFunction(const GraphFunction *function, const CompilationOptions& options) {
```
const GraphFunction&

qihqi replied:
Can you make jitModuleToMobile also calls this function? so we have less code dup

tugsbayasgalan replied:
sure

qihqi commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -248,7 +252,54 @@ std::unique_ptr<mobile::Code> compileGraphToMobileCode(
+  return mobile_func;
+}
+
+IValue convertMobileFunctionToCodeTable(const mobile::Function *func, const CompilationOptions& compilation_options){
```
const &

qihqi replied:
space before {

tugsbayasgalan replied:
sounds good

tugsbayasgalan committed:
Update on "Expose bytecode to Python"

In this PR, we expose upgrader bytecode to mobile through pybind. 
1. We convert TS graph of upgrader to GraphFunction
2. We call compileJitFunctionToMobileFunction to convert GraphFunction to mobile function.
3. We convert the mobile function to a code table of IValues (note that this should be changed in the future when we migrate away from Ivalue based storage) 

Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/serialization/export_bytecode.h:
```diff
@@ -27,5 +27,19 @@ TORCH_API mobile::Module jitModuleToMobile(
+    const CompilationOptions& options);
+
+TORCH_API IValue convertMobileFunctionToCodeTable(
+    const mobile::Function* func,
```
const ref vs raw ptr

gmagogsfm commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -213,7 +213,6 @@ std::pair<IValue, IValue> getFunctionTuple(
         `TypeNameUniquer` to get the managled name of the argument. This helps
         in having the right object reference when a class method is called using
         the `self` argument.
-
```
accidental change?

qihqi commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1593,6 +1599,15 @@ void initJitScriptBindings(PyObject* module) {
+      "_compile_graph_to_code_table",
+      [](const std::string& name, const std::shared_ptr<Graph>& graph) {
+        CompilationOptions options;
+        auto jitFunc = std::make_unique<GraphFunction>(name, graph, nullptr);
```
nit: can probably allocate on stack instead. i.e. `GraphFunction jitFunction(name, graph, nullptr);`

cccclai commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -248,7 +247,60 @@ std::unique_ptr<mobile::Code> compileGraphToMobileCode(
+  return mobile_func;
+}
+
+IValue convertMobileFunctionToCodeTable(
```
It will be made a mobile function again during mobile runtime build. mobile build is a separate from desktop build and it doesn't include `export_bytecode.cpp`. It will include a file like `upgrader_mobile.h` generated from desktop build, which includes a map (key: upgrader name, value bytecode)

As for context, desktop build is `setup.py` and mobile build is like `build_ios.sh`

tugsbayasgalan replied:
Another option is to just create a table of strings and integers as a bytecode. So we can bypass the converting to IValue part.

tugsbayasgalan replied:
Actually I feel like creating a table of strings and integers are bit ugly and hard to maintain. Because you would have to create a custom structs to represent them and expose them through pybind. (instruction for example is a tuple of (str, int, int)) Would creating IValue create that much overhead given that this bytecode generation only used during build process? @gmagogsfm @cccclai

gmagogsfm replied:
It wouldn't be that much overhead, just that C++ structures would be cleaner (well defined struct vs generic IValue). Maybe let's leave this as a TODO for future cleanup?

cccclai replied:
Maybe leave it as is and swap to flatbuffer when it's ready?

cccclai reviewed (approved):
LGTM from edge side. Would like to see more comments when exporting bytecode, ideally throw exception for NamedTuple.

tugsbayasgalan committed:
Update on "Expose bytecode to Python"

In this PR, we expose upgrader bytecode to mobile through pybind. 
1. We convert TS graph of upgrader to GraphFunction
2. We call compileJitFunctionToMobileFunction to convert GraphFunction to mobile function.
3. We convert the mobile function to a code table of IValues (note that this should be changed in the future when we migrate away from Ivalue based storage) 

Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Expose bytecode to Python"

In this PR, we expose upgrader bytecode to mobile through pybind. 
1. We convert TS graph of upgrader to GraphFunction
2. We call compileJitFunctionToMobileFunction to convert GraphFunction to mobile function.
3. We convert the mobile function to a code table of IValues (note that this should be changed in the future when we migrate away from Ivalue based storage) 

Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Expose bytecode to Python"

In this PR, we expose upgrader bytecode to mobile through pybind. 
1. We convert TS graph of upgrader to GraphFunction
2. We call compileJitFunctionToMobileFunction to convert GraphFunction to mobile function.
3. We convert the mobile function to a code table of IValues (note that this should be changed in the future when we migrate away from Ivalue based storage) 

Differential Revision: [D31632600](https://our.internmc.facebook.com/intern/diff/D31632600)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632600).
